pydra/engine/specs.py

```diff
@@ -25,13 +25,11 @@ def hash(self):
         inp_dict = {
             field.name: getattr(self, field.name)
             for field in dc.fields(self)
-            if field.name not in ["_graph"]
+            if field.name not in ["_graph_checksums"]
         }
         inp_hash = hash_function(inp_dict)
-        if hasattr(self, "_graph"):
-            # calculating checksum for every node inside the graph
-            graph_hash = [nd.checksum for nd in self._graph]
-            return hash_function((inp_hash, graph_hash))
+        if hasattr(self, "_graph_checksums"):
+            return hash_function((inp_hash, self._graph_checksums))
```

Comment:
i would compute the graph checksum at this step.

Reply:
but that was in the current master...

Reply:
in the current master `graph_hash = [nd.checksum for nd in self._graph]` this computes the hash for every node at the point of checksum computation, not in the `add` function.

Reply:
in that case I don't know how to remove deepcopy. Nodes are changing, so if we don't create copy and we calculate whenever we want, than the checksum will be changing during execution, and we don't want it.